fabric-component-access-api-v1/src/main/java/net/fabricmc/fabric/impl/component/access/BlockComponentContextImpl.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.component.access;
+
+import net.minecraft.block.Block;
+import net.minecraft.block.BlockState;
+import net.minecraft.block.entity.BlockEntity;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+
+import net.fabricmc.fabric.api.component.access.v1.BlockComponentContext;
+
+@SuppressWarnings("rawtypes")
+public final class BlockComponentContextImpl extends AbstractComponentContextImpl implements BlockComponentContext {
+	private Block block;
+	private BlockPos pos;
+	private BlockState blockState;
+	private BlockEntity blockEntity;
+
+	@Override
+	public Block block() {
+		return block;
+	}
+
+	@Override
+	public BlockPos pos() {
+		BlockPos result = pos;
+
+		if (result == null && blockEntity != null) {
+			result = blockEntity.getPos();
+			pos = result;
+		}
+
+		return result;
+	}
+
+	@Override
+	public BlockEntity blockEntity() {
+		BlockEntity result = blockEntity;
+
+		if (result == null && world != null) {
+			result = world.getBlockEntity(pos);
+			blockEntity = result;
+		}
+
+		return result;
+	}
+
+	@Override
+	public BlockState blockState() {
+		return blockState;
+	}
+
+	@Override
+	protected World getWorldLazily() {
+		return blockEntity == null ? null : blockEntity.getWorld();
```

Comment:
If there is no block entity present at the location, the `this.world` field is entirely ignored.

Reply:
It's ignored by the context but the interface guarantees it will always be available for the provider/component.